Classes/Image Upload/ImageUploadViewController.swift

```diff
@@ -0,0 +1,214 @@
+        // Set the right bar item back to the spinner so they can't spam the button, and to indicate something is happening
+        setRightBarItemSpinning()
+        
+        // Should never caught in here as the button will be disabled in this situation
+        guard let compressionData = compressionData else {
+            StatusBar.showGenericError()
+            return
+        }
+        
+        // Check that we have enough "tokens" to actually upload the image
+        client.canUploadImage { [weak self] success in
+            // Ensure that we do have enough tokens, otherwise remove the upload button
+            guard success else {
+                StatusBar.showError(message: NSLocalizedString("Rate Limit reached, cannot upload!", comment: ""))
+                self?.navigationItem.rightBarButtonItem = nil
+                return
+            }
+            
+            var name = "GitHawk Upload"
+            
+            if let username = self?.username {
+                name += " by \(username)"
+            }
+            
+            // Ensure the upload step is on the background thread
+            DispatchQueue.global(qos: .userInitiated).async {
+                self?.client.uploadImage(
+                    base64: compressionData,
+                    name: name,
+                    title: self?.titleText ?? "",
+                    description: self?.descriptionText ?? "") { [weak self] result in
+                        
+                    // UI Work, so ensure it's on the main thread
+                    DispatchQueue.main.async {
+                        switch result {
+                        case .error:
+                            StatusBar.showGenericError()
+                            self?.setRightBarItemIdle()
+                            return
+                        case .success(let link):
+                            self?.delegate?.imageUploaded(link: link, altText: name)
+                            self?.dismiss(animated: true, completion: nil)
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    // MARK: UITextFieldDelegate
+    
+    /// Called when the user taps return on the title field, moves their cursor to the body
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        bodyTextField.becomeFirstResponder()
+        return false
+    }
+    
+    // MARK: Image Preparation
+    
+    /// Compressed and Encodes in Base64 the provided UIImage.
+    ///
+    /// Process is moved to a background thread in order to prevent UI blocking.
+    ///
+    /// Compression is a value between 0.0 and 1.0. Lower is smaller file size but worse quality.
+    private func compressAndEncodeImage(_ image: UIImage, compression: CGFloat = 0.2, completion: @escaping (Result<String>) -> Void) {
+        DispatchQueue.global(qos: .background).async {
+            let data = UIImageJPEGRepresentation(image, compression)
```

Comment:
Are you sure this shroud be `UIImageJPEGRepresentation`?

Reply:
As opposed to?

I need to convert a UIImage to a Base64 string, if there's a better way of doing this then shout!

Reply:
I misread your initial comment. I thought it meant to be PNG instead of JPG hence my comment. Sorry!